### PR TITLE
correctly check exclusive range patterns for overlap

### PR DESCRIPTION
**Pull request by oli-obk**

### CODE REVIEW

File: tests/compile-fail/matches.rs [hunk 1]
```diff
@@ -246,11 +247,35 @@ fn overlapping() {
     }
 
     match 42 {
+        2 => println!("2"), //~NOTE overlaps with this
+        0 ... 2 => println!("0 ... 5"), //~ERROR: some ranges overlap
```
Comment (mcarton) [inline tests/compile-fail/matches.rs]: The string is wrong here.

File: tests/compile-fail/matches.rs [hunk 2]
```diff
@@ -246,11 +247,35 @@ fn overlapping() {
...
 
+    match 42 {
+        2 => println!("2"),
+        0 .. 2 => println!("0 .. 2"),
+        _ => (),
+    }
+
+    match 42 {
+        0 .. 10 => println!("0 ... 10"),
+        10 .. 50 => println!("0 ... 10"),
```
Comment (mcarton) [inline tests/compile-fail/matches.rs]: Same.

File: tests/compile-fail/matches.rs [hunk 3]
```diff
@@ -246,11 +247,35 @@ fn overlapping() {
...
+        _ => (),
+    }
+
+    match 42 {
+        0 .. 11 => println!("0 ... 10"), //~ERROR: some ranges overlap
```
Comment (mcarton) [inline tests/compile-fail/matches.rs]: Same.

File: tests/compile-fail/matches.rs [hunk 4]
```diff
@@ -246,11 +247,35 @@ fn overlapping() {
...
+        0 ... 11 => println!("0 ... 10"), //~NOTE overlaps with this
```
Comment (mcarton) [inline tests/compile-fail/matches.rs]: Same.

File: clippy_lints/src/matches.rs [hunk 5]
```diff
@@ -470,7 +488,24 @@ pub fn overlapping<T>(ranges: &[SpannedRange<T>]) -> Option<(&SpannedRange<T>, &
 
     impl<'a, T: Copy + Ord> Ord for Kind<'a, T> {
         fn cmp(&self, other: &Self) -> Ordering {
-            self.value().cmp(&other.value())
+            match (self.value(), other.value()) {
+                (Bound::Included(a), Bound::Included(b)) |
+                (Bound::Excluded(a), Bound::Excluded(b)) => a.cmp(&b),
+                (Bound::Unbounded, _) |
+                (_, Bound::Unbounded) => unimplemented!(),
```
Comment (mcarton) [inline clippy_lints/src/matches.rs]: Maybe add a comment that range patterns need to be bounded? I did not know that and had to check.

File: tests/compile-fail/matches.rs (hunk 3 above)
Comment (oli-obk) [inline tests/compile-fail/matches.rs]: it's kind of correct

**Pull request merged**